mesonpy/__init__.py

```diff
@@ -414,11 +414,57 @@ def _is_native(self, file: Union[str, pathlib.Path]) -> bool:
             return True
         return False
 
+    def _add_mesonpy_libs_to_rpath(self, path: pathlib.Path, destination: pathlib.Path) -> pathlib.Path:
+        """Adds .mesonpy.libs to the RPATH """
+        if platform.system() not in ('Linux', 'Darwin'):
+            # Internal libraries are currently unsupported on this platform
+            raise NotImplementedError(
+                'Bundling libraries in wheel is not supported '
+                f"on platform '{platform.system()}'"
+            )
+
+        if not self._is_native(path):
+            return path
+
+        mesonpy_libs_path = os.path.relpath(
+            f'.{self._project.name}.mesonpy.libs',
+            destination.parent,
+        )
+
+        binary: Union[None, mesonpy._elf.ELF, mesonpy._dylib.Dylib] = None
+        libdir_path: Optional[str] = None
+        if platform.system() == 'Linux':
+            binary = mesonpy._elf.ELF(path)
+            libdir_path = f'$ORIGIN/{mesonpy_libs_path}'
+        elif platform.system() == 'Darwin':
+            binary = mesonpy._dylib.Dylib(path)
+            libdir_path = f'@loader_path/{mesonpy_libs_path}'
+
+        # add our .mesonpy.libs directory to RPATH
+        assert binary and libdir_path
+        if libdir_path not in binary.rpath:
+            binary.rpath = [*binary.rpath, libdir_path]
+
+        # fix LC_LOAD_DYLIB entries on macOS to use the RPATH
+        if platform.system() == 'Darwin':
+            libs = {
+                destination.name: destination
+                for destination, _ in self._wheel_files['mesonpy-libs']
+            }
+
+            assert isinstance(binary, mesonpy._dylib.Dylib)
+            for entry in binary.load_dylib:
+                libname = entry.split('/')[-1]
+                if libname in libs:
+                    binary.replace_load_dylib(entry, f'@rpath/{libs[libname].as_posix()}')
```

Comment:
Uhm. I need to research this a bit more, but the impression is that there is something strange here. I'm not sure how to reconcile the different observations:

1. Meson adds an rpath during compilation pointing to the build directory, and that is sufficient to make things work,
2. if we move the libraries to  the `.mesonpy.libs` folder, add one rpath that we thing points to that folder and leave the rpath added by Meson in place (because we pick the files from the build directory and not from the install directory) things keep working,
3. if we do the same but we strip the rpath added by Meson (by picking the files from the install directory) things stop working

Unless there is a funny bug somewhere, I don't think that removing one rpath pointing to the wrong directory in 3. is what breaks things. Looking at what Meson does on install https://github.com/mesonbuild/meson/blob/5dc4fcae34ee3a5a3d47021f8ea8b2c5d3b14ea9/mesonbuild/scripts/depfixer.py#L410-L458 one thing that changes on install in the `-install_name` that gets rewritten from an `@rpath` relative path to an absolute path. May this be what needs to be reverted?

In particular, going from 2. where things work to 3. where things do not work, `LC_LOAD_DYLIB` entries did not change, thus I'm not convinced having to mess with these is the right solution.

I'll investigate more later.

Reply:
> In particular, going from 2. where things work to 3. where things do not work, `LC_LOAD_DYLIB` entries did not change, thus I'm not convinced having to mess with these is the right solution.

The `LC_LOAD_DYLIB` entry did change, and that why it breaks. Feel free to play with it locally.

https://github.com/mesonbuild/meson/blob/5dc4fcae34ee3a5a3d47021f8ea8b2c5d3b14ea9/mesonbuild/scripts/depfixer.py#LL447C49-L447C49